tracee-ebpf/tracee/containers.go

```diff
@@ -0,0 +1,205 @@
+package tracee
+
+import (
+	"bufio"
+	"fmt"
+	"io/fs"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+// Information about running containers
+
+type Containers struct {
+	cgroupv1   bool
+	cgroupv1mp string
+	cgroupv2   bool
+	cgroupv2mp string
+	mapContIds map[string][]string
+}
+
+// Add, Remove, Check, List & Get
+
+func (c *Containers) init() {
+	// (re)initializes Containers
+
+	if c.mapContIds == nil {
+		c.mapContIds = make(map[string][]string, 0)
+	} else {
+		for _, cont := range c.GetContainers() {
+			c.delContainer(cont)
+		}
+		c.mapContIds = make(map[string][]string, 0)
+	}
+
+	c.cgroupv1 = false
+	c.cgroupv2 = false
+	c.cgroupv1mp = ""
+	c.cgroupv2mp = ""
+}
+
+func (c *Containers) addContainer(contId string) {
+	_, ok := c.mapContIds[contId]
+	if !ok {
+		c.mapContIds[contId] = make([]string, 0)
+	}
+}
+
+func (c *Containers) delContainer(contId string) {
+	delete(c.mapContIds, contId) // no-op if no contId
+}
+
+func (c *Containers) checkContainer(contId string) bool {
+	_, ok := c.mapContIds[contId]
+	return ok
+}
+
+func (c *Containers) GetContainers() []string {
+	var conts []string
+	for k := range c.mapContIds {
+		conts = append(conts, k)
+	}
+	if len(conts) == 0 {
+		return nil
+	}
+
+	return conts
+}
+
+func (c *Containers) addPid(contId string, pid string) {
+	c.addContainer(contId)
+	c.mapContIds[contId] = append(c.mapContIds[contId], pid)
+}
+
+func (c *Containers) GetPids(contId string) []string {
+	if !c.checkContainer(contId) {
+		return nil
+	}
+	return c.mapContIds[contId]
+}
+
+// Populate info
+
+func (c *Containers) Populate() error {
+	// do all the hard work
+
+	c.init()
+
+	err := c.procMountsCgroups()
+	if err != nil {
+		return err
+	}
+
+	return c.populate()
+}
+
+func (c *Containers) procMountsCgroups() error {
+	// find cgroups v1 and v2 mountpoints for procfs walks
+
+	mountsFile := "/proc/mounts"
+	file, err := os.Open(mountsFile)
+	if err != nil {
+		return err
+	}
+	scanner := bufio.NewScanner(file)
+	for i := 1; scanner.Scan(); i++ {
+		sline := strings.Split(scanner.Text(), " ")
+		mountpoint := sline[1]
+		fstype := sline[2]
+		if fstype == "cgroup" {
+			if strings.Contains(mountpoint, "cgroup/pids") {
+				c.cgroupv1 = true
+				c.cgroupv1mp = mountpoint
+			}
+		} else if fstype == "cgroup2" {
+			c.cgroupv2 = true
+			c.cgroupv2mp = mountpoint
+		}
+	}
+	_ = file.Close()
+
+	return nil
+}
+
+func (c *Containers) populate() error {
+	// Prepare regex to be used for population done in proc walk
+
+	if c.cgroupv1 {
+		var r []string
+		// docker:  <cgroupv1mp>/<random>/docker/<id>/tasks
+		r = append(r, ".*docker.*(.{64})/tasks$")
+		// podman:  <cgroupv1mp>/<random>/libpod-<id>.scope/tasks
+		r = append(r, ".*libpod.*(.{64})\\.scope/tasks$")
```

Comment:
Are these needed now that you have "generic" regex?

Reply:
I have not tested the generic regexes while I have tested the other 4. I want to make sure the 4 are good and the 3rd generic regex is a 'safe measure'. We can change later if redundant.